### PR TITLE
Update re-attached widget removes to use detach

### LINKED ISSUE #1902: Support `movableContentOf` across protocol
Currently, due to #1899, we crash when the detached subtree is re-added because children IDs clash. Even once that's fixed, we need to detect the fact that it's a reused node (probably with something like a `removed=true` boolean on the node) and upon re-insertion, walk its children and add them all to the map.

Once guest-side is fixed, we basically have the same problem host-side. It has the map leak that will cause the crash (#1900), but once that's fixed we need the same tree walk on re-attach.

It is unlikely that we will retain the host-side nodes across recompositions. If you are just moving the content in a single composition, however, we can send a new boolean in the remove op to signal that this is a detach with a future op performing re-attach via insertion. This will keep all nodes in the map, and only perform a re-parenting from the old children to the new children.

### PR DESCRIPTION
**Pull request by JakeWharton**
This ensures that the host-side maintains their instances to correctly support re-attach.

Closes #1902 

[movable.webm](https://github.com/user-attachments/assets/744ccb40-0f9c-40e9-acaf-afe8b6684056)

---

- [x] `CHANGELOG.md`'s "Unreleased" section has been updated, if applicable.

### CODE REVIEW

**Review by colinrtwhite — APPROVED** [review]
Nice!

File: redwood-protocol-guest/src/commonMain/kotlin/app/cash/redwood/protocol/guest/DefaultGuestProtocolAdapter.kt [hunk 1]
```diff
@@ -136,9 +137,18 @@ public class DefaultGuestProtocolAdapter(
...
-      "Attempted to add widget with ID ${child.id} but one already exists"
+    val childId = child.id.value
+    if (child.removeIndex != INVALID_INDEX) {
+      check(hostSupportsRemoveDetach) { "Host v$hostVersion does not support widget re-attach" }
+      check(childId in widgets) { "Attempted to re-attach unknown widget with ID $childId" }
+      removed.remove(childId)
+
+      // Update the remove change to indicate it's only a detach.
+      val remove = changes[child.removeIndex] as ChildrenChange.Remove
+      changes[child.removeIndex] = ChildrenChange.Remove(remove.id, remove.tag, remove.index, true)
```
Comment (squarejesse) [inline redwood-protocol-guest/src/commonMain/kotlin/app/cash/redwood/protocol/guest/DefaultGuestProtocolAdapter.kt]: name the parameter `true` ?

File: redwood-protocol-guest/src/commonTest/kotlin/app/cash/redwood/protocol/guest/ProtocolTest.kt [hunk 2]
```diff
@@ -288,6 +295,171 @@ class ProtocolTest {
...
+          }
+        }
+      }
+      when (state) {
+        0 -> Row { oneTwoThree() }
+        1 -> Column { oneTwoThree() }
+      }
+    }
+
+    assertThat(composition.awaitSnapshot().filter { it !is ValueChange }).containsExactly(
```
Comment (squarejesse) [inline redwood-protocol-guest/src/commonTest/kotlin/app/cash/redwood/protocol/guest/ProtocolTest.kt:318]: Beautiful!

File: redwood-protocol-guest/src/commonTest/kotlin/app/cash/redwood/protocol/guest/ProtocolTest.kt [hunk 3]
```diff
@@ -288,6 +295,171 @@ class ProtocolTest {
...
+      ChildrenChange.Add(Id(2), ChildrenTag(1), Id(4), 1),
+      // Text("three") to Box
+      Create(Id(5), WidgetTag(3)),
+      ChildrenChange.Add(Id(2), ChildrenTag(1), Id(5), 2),
+      // Box to Row
+      ChildrenChange.Add(Id(1), ChildrenTag(1), Id(2), 0),
+    )
+
+    state = 1
+    assertThat(composition.awaitSnapshot().filter { it !is ValueChange }).containsExactly(
```
Comment (squarejesse) [inline redwood-protocol-guest/src/commonTest/kotlin/app/cash/redwood/protocol/guest/ProtocolTest.kt:338]: The `ValueChange` instances all target the new `Column` here, right?

File: redwood-protocol-guest/src/commonTest/kotlin/app/cash/redwood/protocol/guest/ProtocolTest.kt [hunk 4]
```diff
@@ -288,6 +295,171 @@ class ProtocolTest {
...
+      // Box from Row
+      ChildrenChange.Remove(Id(1), ChildrenTag(1), 0, detach = true),
```
Comment (squarejesse) [inline redwood-protocol-guest/src/commonTest/kotlin/app/cash/redwood/protocol/guest/ProtocolTest.kt:340]: YASS

File: redwood-protocol-guest/src/commonTest/kotlin/app/cash/redwood/protocol/guest/ProtocolTest.kt [hunk 5]
```diff
@@ -288,6 +295,171 @@ class ProtocolTest {
...
+      ChildrenChange.Remove(Id.Root, ChildrenTag.Root, 0, detach = false),
+      // Column to Root
+      Create(Id(6), WidgetTag(1000002)),
+      ChildrenChange.Add(Id.Root, ChildrenTag.Root, Id(6), 0),
+      // Box to Column
+      ChildrenChange.Add(Id(6), ChildrenTag(1), Id(2), 0),
+    )
+  }
+
+  @Test fun multipleMovableContentButOnlyOneReused() = runTest {
```
Comment (squarejesse) [inline redwood-protocol-guest/src/commonTest/kotlin/app/cash/redwood/protocol/guest/ProtocolTest.kt:351]: So cool

**Review by squarejesse — APPROVED** [review]

**Comment by squarejesse** [comment]
You made it look easy

File: redwood-tooling-codegen/src/main/kotlin/app/cash/redwood/tooling/codegen/protocolGuestGeneration.kt [hunk 6]
```diff
@@ -194,6 +194,7 @@ internal class ProtocolButton(
 ) : ProtocolWidget, Button<Unit> {
   public override val id: Id = guestAdapter.nextId()
   public override val tag: WidgetTag get() = WidgetTag(3)
+  public override var removeIndex: Int = InvalidRemoveIndex
```
Comment (JakeWharton) [inline redwood-tooling-codegen/src/main/kotlin/app/cash/redwood/tooling/codegen/protocolGuestGeneration.kt]:
```suggestion
  public override var removeIndex: Int = INVALID_INDEX
```

File: redwood-protocol-guest/src/commonMain/kotlin/app/cash/redwood/protocol/guest/DefaultGuestProtocolAdapter.kt (hunk 1 above)
Comment (JakeWharton) [inline redwood-protocol-guest/src/commonMain/kotlin/app/cash/redwood/protocol/guest/DefaultGuestProtocolAdapter.kt]:
```suggestion
      changes[child.removeIndex] = ChildrenChange.Remove(remove.id, remove.tag, remove.index, detach = true)
```

**Comment by JakeWharton** [comment]
> You made it look easy

Extracting like 8 infrastructure PRs out from before it certainly helped!

**Pull request merged**